crates/util/edcode-derive/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+//! Proc-macros for deriving `rimecraft_edcode` traits.
+//!
+//! __You shouldn't use this crate directly__, use `rimecraft_edcode` crate
+//! with `derive` feature flag instead.
+
+use proc_macro::TokenStream;
+use proc_macro2::TokenTree;
+use quote::quote;
+use syn::{
+    parse_macro_input, spanned::Spanned, Attribute, Data, DataEnum, DeriveInput, Error, Expr,
+    Fields, Ident, Meta,
+};
+
+macro_rules! unsupported_object {
+    ($tr:literal, $ty:literal) => {
+        concat!("deriving `", $tr, "` to `", $ty, "` is not supported")
+    };
+}
+
+macro_rules! fields_disallowed {
+    () => {
+        "variants with fields are not supported"
+    };
+}
+
+macro_rules! discriminant_required {
+    () => {
+        "variants must have explicit discriminant"
+    };
+}
+
+macro_rules! unsupported_repr {
+    () => {
+        "only `u*` and `i*` (excluding 128-bit types) repr is supported"
+    };
+}
+
+macro_rules! repr_required {
+    () => {
+        "must specify repr"
+    };
+}
+
+/// Common parsing code for deriving to `enum`.
+fn parse_derive_enum(
+    ident: Ident,
+    attrs: Vec<Attribute>,
+    data: DataEnum,
+) -> Result<(Ident, Ident, Vec<Ident>, Vec<Expr>), TokenStream> {
+    let mut enum_idents: Vec<Ident> = vec![];
+    let mut enum_vals: Vec<Expr> = vec![];
+    for var in data.variants.into_iter() {
+        if let Fields::Unit = var.fields {
+            enum_idents.push(var.ident.clone());
+        } else {
+            return Err(Error::new(var.fields.span(), fields_disallowed!())
+                .into_compile_error()
+                .into());
+        }
+        let has_disc_err = !var.discriminant.clone().is_some_and(|(_, e)| {
```

Comment:
`is_some_and` accepts a predicate and a predicate shouldn't change the environment, I think.

Reply:
Although technically it works as it only accepts `FnOnce`.